AlgorithmFactory/Loader.cs

```diff
@@ -140,80 +140,88 @@ public bool TryCreateAlgorithmInstance(string assemblyPath, out IAlgorithm algor
-                var items = new List<KeyValuePair<string, dynamic>>();
-                try
+                // Initialize Python Engine
+                if (!PythonEngine.IsInitialized)
                 {
-                    Log.Trace("Loader.TryCreatePythonAlgorithm(): Importing python module...");
-                    var algorithmName = Config.Get("algorithm-type-name");
-                    var scope = engine.Runtime.ImportModule(string.IsNullOrEmpty(algorithmName) ? "main" : algorithmName);
-                    items = (List<KeyValuePair<string, dynamic>>)scope.GetItems();
+                    PythonEngine.Initialize();
+                    PythonEngine.BeginAllowThreads();
                 }
-                catch (Exception err)
+                
+                using (Py.GIL())
                 {
-                    Log.Error(err);
-                    errorMessage = err.Message + " - could not locate 'main' module. Please make sure you have a main.py file in your project.";
-                    return false;
-                }
+                    Log.Trace("Loader.TryCreatePythonAlgorithm(): Importing python module " + algorithmName);
 
-                //Loop through the types in the dll, see if we can find a "QCAlgorithm" base class
-                Log.Trace("Loader.TryCreatePythonAlgorithm(): Finding QCAlgorithm...");
-                dynamic dynamicAlgorithm = null;
-                foreach (var item in items)
-                {
-                    try
+                    var scope = Py.Import(algorithmName);
+                    if (scope == null)
                     {
-                        string baseName = item.Value.__bases__.ToString().ToString();
-                        if (baseName.Contains("QCAlgorithm"))
+                        errorMessage = "Loader.TryCreatePythonAlgorithm(): Unable to import python module. Check for errors in the py file.";
+                        return false;
+                    }
+
+                    var baseClass = scope.GetAttr("QCAlgorithm");
+                    if (baseClass == null)
+                    {
+                        errorMessage = "Loader.TryCreatePythonAlgorithm(): Unable to find QCAlgorithm class in module.";
+                        return false;
+                    }
+
+                    Log.Trace("Loader.TryCreatePythonAlgorithm(): Creating IAlgorithm instance.");
+                    
+                    foreach (var name in scope.Dir())
+                    {
+                        var attr = scope.GetAttr(name.ToString());
+
+                        if (attr.IsSubclass(baseClass) && attr.Repr().Contains(algorithmName))
                         {
-                            dynamicAlgorithm = item.Value;
+                            var addUniverse = Py.Import("qc_algorithm_helper").GetAttr("AddUniverse");
+                            attr.SetAttr("Add_Universe", addUniverse);
```

Comment:
If anyone has ideas how to remove this we'd like to ideally keep the "AddUniverse" method name but override its implementation with our own helper method. This underscore is a temporary work around but will cause confusion if its pushed to production.